Upgrade Hive to 4.x, Hadoop to 3.4.x and Java to 11/17/21

adamdec0173 opened the pull request:
Code to add support for:
- Jdk 11/17/21
- hive-jdbc 4.0.1
- hadoop 3.4.1

massdosage commented:
This MR contains a lot of changes that don't seem to be related to changing the versions. Are you sure you started off with the latest code from the `main` branch here? There are lots of changes in here which seem to mix the order of imports, change the order of methods, add copyright years to license headers etc. which make it much harder to review the bare minimum change. Can you please remove any changes that aren't directly related to the version bumps?

adamdec0173 commented:
IntelliJ has had to format the java code....I can try to revert but this will now take a lot of time

adamdec0173 commented:
Reverted code formatting.

Fixed unit tests.

Only WARN in logs for now is:

```
2024-10-16T20:46:13,259 WARN  DataNucleus.Datastore.Schema:83 - The MetaData for "org.apache.hadoop.hive.metastore.model.MStorageDescriptor" is specified with a foreign-key at class level yet no "table" is defined. All foreign-keys at this level must have a table that the FK goes to.
2024-10-16T20:46:13,668 WARN  org.apache.hadoop.hive.metastore.MetastoreDirectSqlUtils:79 - Failed to execute [select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ? ] with parameters [hive]
javax.jdo.JDODataStoreException: Error executing SQL query "select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ?".
	at org.datanucleus.api.jdo.JDOAdapter.getJDOExceptionForNucleusException(JDOAdapter.java:608) ~[datanucleus-api-jdo-6.0.3.jar:?]
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:456) ~[datanucleus-api-jdo-6.0.3.jar:?]
	at org.datanucleus.api.jdo.JDOQuery.executeWithArray(JDOQuery.java:318) ~[datanucleus-api-jdo-6.0.3.jar:?]
	at org.apache.hadoop.hive.metastore.MetastoreDirectSqlUtils.executeWithArray(MetastoreDirectSqlUtils.java:69) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.executeWithArray(MetaStoreDirectSql.java:2397) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.executeWithArray(MetaStoreDirectSql.java:2392) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.getFunctionIds(MetaStoreDirectSql.java:3327) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.getFunctions(MetaStoreDirectSql.java:3241) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.ObjectStore$30.getSqlResult(ObjectStore.java:11818) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.ObjectStore$30.getSqlResult(ObjectStore.java:11815) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.ObjectStore$GetHelper.run(ObjectStore.java:4490) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.ObjectStore.getFunctionsInternal(ObjectStore.java:11829) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.ObjectStore.getAllFunctions(ObjectStore.java:11807) ~[hive-exec-4.0.1.jar:4.0.1]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.apache.hadoop.hive.metastore.RawStoreProxy.invoke(RawStoreProxy.java:97) ~[hive-exec-4.0.1.jar:4.0.1]
	at jdk.proxy2.$Proxy33.getAllFunctions(Unknown Source) ~[?:?]
	at org.apache.hadoop.hive.metastore.HMSHandler.get_all_functions(HMSHandler.java:8937) ~[hive-exec-4.0.1.jar:4.0.1]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:580) ~[?:?]
	at org.apache.hadoop.hive.metastore.RetryingHMSHandler.invokeInternal(RetryingHMSHandler.java:91) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.metastore.AbstractHMSHandlerProxy.invoke(AbstractHMSHandlerProxy.java:82) ~[hive-exec-4.0.1.jar:4.0.1]
	at jdk.proxy2.$Proxy35.get_all_functions(Unknown Source) ~[?:?]
	at org.apache.hadoop.hive.metastore.HiveMetaStoreClient.getAllFunctions(HiveMetaStoreClient.java:4781) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.getAllFunctions(Hive.java:6221) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.reloadFunctions(Hive.java:362) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.registerAllFunctionsOnce(Hive.java:341) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.<init>(Hive.java:567) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.create(Hive.java:459) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.getInternal(Hive.java:446) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.get(Hive.java:531) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.metadata.Hive.get(Hive.java:520) ~[hive-exec-4.0.1.jar:4.0.1]
	at org.apache.hadoop.hive.ql.plan.mapper.MetastoreStatsConnector$RuntimeStatsLoader.run(MetastoreStatsConnector.java:85) ~[hive-exec-4.0.1.jar:4.0.1]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572) ~[?:?]
	at java.util.concurrent.FutureTask.run(FutureTask.java:317) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642) ~[?:?]
	at java.lang.Thread.run(Thread.java:1583) ~[?:?]
Caused by: java.sql.SQLSyntaxErrorException: Table/View 'FUNCS' does not exist.
	at org.apache.derby.impl.jdbc.SQLExceptionFactory.getSQLException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.Util.generateCsSQLException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.TransactionResourceImpl.wrapInSQLException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.TransactionResourceImpl.handleException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedConnection.handleException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.ConnectionChild.handleException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedPreparedStatement.<init>(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedPreparedStatement42.<init>(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.jdbc.Driver42.newEmbedPreparedStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedConnection.prepareStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedConnection.prepareStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.datanucleus.store.rdbms.SQLController.getStatementForQuery(SQLController.java:353) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.RDBMSQueryUtils.getPreparedStatementForQuery(RDBMSQueryUtils.java:224) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.SQLQuery.performExecute(SQLQuery.java:657) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.query.Query.executeQuery(Query.java:2004) ~[datanucleus-core-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.SQLQuery.executeWithArray(SQLQuery.java:826) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:433) ~[datanucleus-api-jdo-6.0.3.jar:?]
	... 36 more
Caused by: org.apache.derby.iapi.error.StandardException: Table/View 'FUNCS' does not exist.
	at org.apache.derby.iapi.error.StandardException.newException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.iapi.error.StandardException.newException(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.FromBaseTable.bindTableDescriptor(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.FromBaseTable.bindNonVTITables(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.TableOperatorNode.bindNonVTITables(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.FromList.bindTables(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.SelectNode.bindNonVTITables(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.DMLStatementNode.bindTables(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.DMLStatementNode.bind(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.compile.CursorNode.bindStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.GenericStatement.prepMinion(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.GenericStatement.prepare(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.sql.conn.GenericLanguageConnectionContext.prepareInternalStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedPreparedStatement.<init>(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedPreparedStatement42.<init>(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.jdbc.Driver42.newEmbedPreparedStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedConnection.prepareStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.apache.derby.impl.jdbc.EmbedConnection.prepareStatement(Unknown Source) ~[derby-10.14.2.0.jar:?]
	at org.datanucleus.store.rdbms.SQLController.getStatementForQuery(SQLController.java:353) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.RDBMSQueryUtils.getPreparedStatementForQuery(RDBMSQueryUtils.java:224) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.SQLQuery.performExecute(SQLQuery.java:657) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.store.query.Query.executeQuery(Query.java:2004) ~[datanucleus-core-6.0.8.jar:?]
	at org.datanucleus.store.rdbms.query.SQLQuery.executeWithArray(SQLQuery.java:826) ~[datanucleus-rdbms-6.0.8.jar:?]
	at org.datanucleus.api.jdo.JDOQuery.executeInternal(JDOQuery.java:433) ~[datanucleus-api-jdo-6.0.3.jar:?]
	... 36 more
2024-10-16T20:46:13,676 WARN  org.apache.hadoop.hive.metastore.ObjectStore:4533 - Falling back to ORM path due to direct SQL failure (this is not an error): See previous errors; Error executing SQL query "select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ?".Failed to execute [select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ? ] with parameters [hive] at org.apache.hadoop.hive.metastore.MetastoreDirectSqlUtils.executeWithArray(MetastoreDirectSqlUtils.java:81) at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.executeWithArray(MetaStoreDirectSql.java:2397) at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.executeWithArray(MetaStoreDirectSql.java:2392) at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.getFunctionIds(MetaStoreDirectSql.java:3327) at org.apache.hadoop.hive.metastore.MetaStoreDirectSql.getFunctions(MetaStoreDirectSql.java:3241) at org.apache.hadoop.hive.metastore.ObjectStore$30.getSqlResult(ObjectStore.java:11818)
```

adamdec0173 commented on src/test/resources/AggregateViewTest/create_table.sql:
```diff
@@ -1,6 +1,6 @@
 CREATE DATABASE db;
 
-CREATE EXTERNAL TABLE `db.mvtdescriptionchangeinfo`(
```
Throws exception with `

adamdec0173 commented on pom.xml:
```diff
@@ -1,360 +1,605 @@
+                </configuration>
+            </plugin>
+
+<!--            <plugin>-->
```
Commented as it changes java file license on every build

massdosage commented on pom.xml:
```diff
@@ -1,360 +1,605 @@
-        </repository>
-      </distributionManagement>
-      <build>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
Please revert whatever formatting changes are causing this diff to consider the entire file as changed.

massdosage commented:
There are still lots of changes here which don't appear to be directly related to the version bumps, can you please revert things like changes to the order of imports etc. We need to get this down to the bare minimum of changes otherwise there are just too many differences to review.

adamdec0173 commented:
> There are still lots of changes here which don't appear to be directly related to the version bumps, can you please revert things like changes to the order of imports etc. We need to get this down to the bare minimum of changes otherwise there are just too many differences to review.

I did my best, essentially the whole project should be formatted. I have no clue from where this empty lines are coming from. I have UTF-8 encoding.

massdosage commented on pom.xml:
```diff
@@ -33,154 +33,397 @@
+        <junit.jupiter.version>5.11.2</junit.jupiter.version>
+        <junit.version>4.13.2</junit.version>
+        <tez.version>0.10.4</tez.version>
+        <mockito.version>5.14.2</mockito.version>
```
Can you put all the versions that aren't used more than once back down into the `dependency` declaration? Up to now the only ones declared as properties are the ones which are repeated.

adamdec0173 replied:
I can do that bu it the same way this was added before my changes:

<license.maven.plugin.version>3.0</license.maven.plugin.version>

so I do not see a one approach here/

adamdec0173 replied:
Done.

massdosage commented on pom.xml:
```diff
@@ -33,154 +33,397 @@
+    <properties>
+        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+        <hadoop.version>3.4.0</hadoop.version>
+        <hive.version>4.0.1</hive.version>
```
We're going to have to think about how we handle this as it's obviously a backwards-incompatible change. What we did in the past with `hive-3` was we put that into a separate branch and kept `main` for hive 2. At some point we decided Hive 3 had reached critical mass and we then swapped them around. So we'll probably want to do something similar here, we'll have a think. For now we can continue looking at the PR against `main` but we might want to move this over to a different branch.

adamdec0173 replied:
Sure, thanks. I would also like to fix here some other issues as well. But lets focus on this PR first.

massdosage commented on pom.xml:
```diff
@@ -199,16 +442,16 @@
         <configuration>
-          <source>1.8</source>
-          <target>1.8</target>
+          <source>21</source>
+          <target>21</target>
```
We don't want to force 21. I'm happy with moving to Java 11 instead of 8 but I don't want to force all users to move to Java 21. The code should work on >=11.

adamdec0173 replied:
Done.

massdosage commented:
Can you change the destination branch for this PR from `main` to `hive-4.x` which I have just created. An option would be to release this with a major version bump to `7` for Hive 4 and we keep `6.x` versions for Hive 3. At some point in the future if/when Hive 4 becomes the main version of Hive we can create a `hive-3.x`branch for legacy Hive 3 like we do for Hive 2 with `hive-2.x` and then we can swap `hive-4.x` and `main` around.

adamdec0173 commented:
Hi, I see it is already  hive-4.x


Confidential


Confidential

________________________________
From: Adrian Woodhead ***@***.***>
Sent: Thursday, 24 October 2024 15:09
To: HiveRunner/HiveRunner ***@***.***>
Cc: Dec, Adam ***@***.***>; Author ***@***.***>
Subject: Re: [HiveRunner/HiveRunner] Upgrade Hive to 4.x, Hadoop to 3.4.x and Java to 11 (PR #181)


CAUTION: This email originated from outside of Nordea. Please be careful when clicking links or opening attachments.

Can you change the destination branch for this PR from main to hive-4.x which I have just created. An option would be to release this with a major version bump to 7 for Hive 4 and we keep 6.x versions for Hive 3. At some point in the future if/when Hive 4 becomes the main version of Hive we can create a hive-3.xbranch for legacy Hive 3 like we do for Hive 2 with hive-2.x and then we can swap hive-4.x and main around.

—
Reply to this email directly, view it on GitHub<https://github.com/HiveRunner/HiveRunner/pull/181#issuecomment-2435255314>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BMETU5RM375CYXN6AMEW2LTZ5DWR3AVCNFSM6AAAAABQBC47CWVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMZVGI2TKMZRGQ>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
 
 The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/) and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
+## [6.2.0] - 2024-10-16
```
```suggestion
## [TBD
```
We don't know the version or the date yet, this will be filled in when we're ready to do a release.

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
 
+## [6.2.0] - 2024-10-16
+### Changed
+- Updated all dependency versions for hive, hadoop, datanucleus, reflections, guava, tez, mockito, junit and added them in pom.xml <properties> section
```
Can you please put the "before" and "after" values like we do in the lines below - "Update foo from `1.2.3` to `2.5.0`" for example.

adamdec0173 replied:
Done

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
+## [6.2.0] - 2024-10-16
+### Changed
+- Updated all dependency versions for hive, hadoop, datanucleus, reflections, guava, tez, mockito, junit and added them in pom.xml <properties> section
+- Updated all maven plugin versions
```
You can leave this without specifying the version or just remove it completely us the end users of HiveRunner shouldn't need to care about what Maven plugins were used to build it.

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
+### Changed
+- Updated all dependency versions for hive, hadoop, datanucleus, reflections, guava, tez, mockito, junit and added them in pom.xml <properties> section
+- Updated all maven plugin versions
+- Adjusted code for Java21 and removed Guava Optional from the code
```
Maybe be more specific with what was changed for Java 11 and put the Guava Optional change on its own line.

adamdec0173 replied:
done

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
+- Updated all dependency versions for hive, hadoop, datanucleus, reflections, guava, tez, mockito, junit and added them in pom.xml <properties> section
+- Updated all maven plugin versions
+- Adjusted code for Java21 and removed Guava Optional from the code
+- Changed Hive property names in StandaloneHiveServerContext
```
Is it useful for the end users to know which ones were changed?

adamdec0173 replied:
No...names are the same..they just added "_"

massdosage commented on CHANGELOG.md:
```diff
@@ -3,6 +3,15 @@ All notable changes to this project will be documented in this file.
+- Adjusted code for Java21 and removed Guava Optional from the code
+- Changed Hive property names in StandaloneHiveServerContext
+- Added missing Hive & Datanucleus properties in StandaloneHiveServerContext so now the framework works on a new hive dependency versions
+- Fixed unit tests
```
This is a given for every release, you don't need to mention it.

adamdec0173 replied:
ok

massdosage commented on README.md:
```diff
@@ -57,22 +57,24 @@ As an alternative if this does not solve the OOM issues, try increase the -Xmx a
-(please note that the forkMode option is deprecated and you should use forkCount and reuseForks instead)
+(please note that the forkMode option is deprecated, you should use forkCount and reuseForks instead)
+
+(please note that -XX:MaxPermSize will not work for Java > 8)
```
Since this PR moves to Java 11 should we just remove the reference to `xPermSize` entirely?

adamdec0173 replied:
done

massdosage commented on README.md:
```diff
@@ -88,15 +90,63 @@ System property `hiveconf_hive.execution.engine` to 'tez'.
+            </configuration>
+        </plugin>
+
+Tun run unit/integration tests, configuration below can be used
```
```suggestion
Tun run unit/integration tests, the configuration below can be used:
```

adamdec0173 replied:
done

massdosage commented on README.md:
```diff
@@ -88,15 +90,63 @@ System property `hiveconf_hive.execution.engine` to 'tez'.
+                    </goals>
+                </execution>
+            </executions>
+            <!-- A hack to run integration tests (*IT.java) if your production code uses org.antlr in version 3 -->
```
What's this about?

adamdec0173 replied:
Added better explanation

massdosage commented on README.md:
```diff
@@ -176,14 +226,14 @@ The [HiveShell](/src/main/java/com/klarna/hiverunner/HiveShell.java) field annot
 # Hive version compatibility
 
-- This version of HiveRunner is built for Hive 3.1.2.
+- This version of HiveRunner is built for Hive 4.0.x.
 - For Hive 2.x support please use HiveRunner 5.x.
```
```suggestion
- This version of HiveRunner is built for Hive 4.0.x.
- For Hive 3.x support please use HiveRunner 6.x.
- For Hive 2.x support please use HiveRunner 5.x.
```

adamdec0173 replied:
done

massdosage reviewed (changes requested):
I just had a first look, will continue with the rest later, probably tomorrow.

massdosage commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+
+        <!-- To make Hive partitions work. Example: SELECT * FROM db.table WHERE partition_column = 'YYZ' -->
+        <dependency>
+            <groupId>org.antlr</groupId>
```
Do we need this here if we can remove the exclusions on Antlr in the Hive dependencies?

adamdec0173 replied:
Yes this is needed otherwise HQL for partitioned EXTERNAL Hive tables  are not working: SELECT * FROM db.table WHERE partition_column = 'YYZ'

massdosage commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+        <version>3.5.1</version>
         <configuration>
-          <!-- Hiverunner need this for some queries (property -XX:MaxPermSize
+          <!-- HiveRunner need this for some queries (property -XX:MaxPermSize
             was removed in java 8) -->
```
```suggestion
          <!-- HiveRunner need this for some queries -->
```

massdosage commented on pom.xml:
```diff
@@ -221,54 +461,54 @@
-        </executions>
-      </plugin>
-      
+<!--      <plugin>-->
```
I'll need to look into this, it should "just work" but I remember your previous set of changes it seemed to be doing something weird.

adamdec0173 replied:
sure, thanks

massdosage commented on pom.xml:
```diff
@@ -330,7 +570,7 @@
             <groupId>org.apache.maven.plugins</groupId>
             <artifactId>maven-gpg-plugin</artifactId>
-            <version>1.6</version>
+            <version>3.2.7</version>
```
I hope this won't introduce any surprises when we try perform an actual release to Maven Central. Something to keep in mind for later.

adamdec0173 replied:
Sure, lets try if not we will revert.

massdosage commented on src/main/java/com/klarna/hiverunner/HiveRunnerExtension.java:
```diff
@@ -74,11 +74,15 @@ public void postProcessTestInstance(Object target, ExtensionContext extensionCon
-        Predicates.and(
-            withAnnotation(HiveRunnerSetup.class),
-            withType(HiveRunnerConfig.class)));
+    private void setupConfig(Object target) {
```
Not sure this is really necessary but I can live with it ;)

adamdec0173 replied:
It is as reflections library is no using GUAVA Predicates anymore rather than java.util.function.Predicate

massdosage commented on src/main/java/com/klarna/hiverunner/StandaloneHiveServerContext.java:
```diff
@@ -16,26 +16,7 @@
-import static org.apache.hadoop.hive.conf.HiveConf.ConfVars.SCRATCHDIR;
-import static org.apache.hadoop.hive.metastore.conf.MetastoreConf.ConfVars.HIVE_IN_TEST;
-
+import static org.apache.hadoop.hive.conf.HiveConf.ConfVars.*;
```
Please expand the * back again.

massdosage commented on src/main/java/com/klarna/hiverunner/StandaloneHiveServerContext.java:
```diff
@@ -195,29 +176,49 @@ protected void configureMetaStore(HiveConf conf) {
-    setMetastoreProperty(METASTORE_VALIDATE_CONSTRAINTS.varname, "true");
-    setMetastoreProperty(METASTORE_VALIDATE_COLUMNS.varname, "true");
-    setMetastoreProperty(METASTORE_VALIDATE_TABLES.varname, "true");
+    * If hive.in.test=false (default), Hive 3 will assume that the metastore rdbms has already been initialized
```
```suggestion
    * If hive.in.test=false (default), Hive will assume that the metastore rdbms has already been initialized
```

adamdec0173 replied:
done

massdosage commented on src/main/java/com/klarna/hiverunner/io/IgnoreClosePrintStream.java:
```diff
@@ -18,11 +18,12 @@
 
     public IgnoreClosePrintStream(OutputStream out) {
-        super(out);
+        super(out, false, StandardCharsets.UTF_8);
```
Can you add a comment here explaining what the `false` is for?

adamdec0173 replied:
reverted

massdosage replied:
So for some reason you needed to disable the `autoFlush`? Or is that the default and there was no other super constructor which allowed you to pass in just the `encoding`?

adamdec0173 replied:
Default was false. Added only UTF-8 as it was throwing NPE.

massdosage commented on src/main/java/com/klarna/reflection/ReflectionUtils.java:
```diff
@@ -113,18 +113,33 @@ private static <T> T getFieldValue(Object testCase, Class testCaseClass, String
     }
-
-}
+}
```
Put EOF closing line break back plz.

adamdec0173 replied:
done

massdosage commented on src/test/java/com/klarna/hiverunner/OrcSnappyTest.java:
```diff
@@ -56,20 +55,20 @@ public void verifyThatDataIsAvailableInOrcSnappy() {
-        List<String> actual = hiveShell.executeQuery("select count(*) from foo_orc_snappy");
-        Assertions.assertEquals(expected, actual);
-    }
+//    @Disabled // Fails with java.lang.UnsatisfiedLinkError: org.apache.hadoop.util.NativeCodeLoader.buildSupportsSnappy()Z
```
If they're disabled we can just keep them? Or just delete them entirely if we think this will never work.

adamdec0173 replied:
Fixed. Works now

massdosage commented on src/test/java/com/klarna/hiverunner/ToUpperCaseSerDe.java:
```diff
@@ -27,18 +27,15 @@
-import java.util.Arrays;
-import java.util.List;
-import java.util.Properties;
+import java.util.*;
```
Expand the `*` back.

massdosage commented on src/test/java/com/klarna/hiverunner/examples/HelloHiveRunnerParameterizedTest.java:
```diff
@@ -16,56 +16,60 @@
+import com.klarna.hiverunner.HiveRunnerExtension;
+import com.klarna.hiverunner.HiveShell;
+import com.klarna.hiverunner.annotations.HiveRunnerSetup;
+import com.klarna.hiverunner.annotations.HiveSQL;
+import com.klarna.hiverunner.config.HiveRunnerConfig;
 import org.junit.jupiter.api.extension.ExtendWith;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.ValueSource;
 
-import com.klarna.hiverunner.HiveRunnerExtension;
-import com.klarna.hiverunner.HiveShell;
-import com.klarna.hiverunner.annotations.HiveSQL;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
```
Why did these get re-ordered?

adamdec0173 replied:
Reverted changes.

massdosage commented on src/test/java/com/klarna/hiverunner/examples/HelloHiveRunnerParameterizedTest.java:
```diff
@@ -16,56 +16,60 @@
+        shell.executeStatement("DROP DATABASE " + dbName + " CASCADE");
     }
-}
+}
```
Closing line break.

adamdec0173 replied:
done

adamdec0173 commented on pom.xml:
```diff
@@ -3,7 +3,7 @@
   <groupId>io.github.hiverunner</groupId>
   <artifactId>hiverunner</artifactId>
-  <version>6.1.1-SNAPSHOT</version>
+  <version>6.2.0-SNAPSHOT</version>
```
done

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+        <dependency>
+            <groupId>org.datanucleus</groupId>
+            <artifactId>datanucleus-api-jdo</artifactId>
+            <version>6.0.3</version>
```
no

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+        </dependency>
+
+        <dependency>
+            <groupId>com.google.guava</groupId>
```
removed, transitive one will be used. IMHO I would redesign the code and remove guava at all. Maybe in next PR?

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>hive-serde</artifactId>
+            <version>${hive.version}</version>
+            <exclusions>
+                <exclusion>
```
Yes, there are a lot of ugly errors, one example: 

java.lang.UnsupportedClassVersionError: org/apache/commons/logging/LogFactory has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>hive-webhcat-java-client</artifactId>
+            <version>${hive.version}</version>
+            <exclusions>
+                <exclusion>
```
They are needed

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>tez-dag</artifactId>
+            <version>${tez.version}</version>
+            <exclusions>
+                <exclusion>
```
They are needed

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>tez-common</artifactId>
+            <groupId>org.apache.tez</groupId>
+            <version>${tez.version}</version>
+            <exclusions>
```
They are needed

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>hadoop-mapreduce-client-common</artifactId>
+            <version>${hadoop.version}</version>
+            <exclusions>
+                <exclusion>
```
They are needed

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <artifactId>hadoop-mapreduce-client-core</artifactId>
+            <version>${hadoop.version}</version>
+            <exclusions>
+                <exclusion>
```
They are needed

adamdec0173 commented on pom.xml:
```diff
@@ -33,155 +33,395 @@
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-client-runtime</artifactId>
+            <version>${hadoop.version}</version>
+            <exclusions>
```
They are needed

adamdec0173 replied:
They are needed

adamdec0173 commented:
2024-10-16T20:46:13,259 WARN  DataNucleus.Datastore.Schema:83 - The MetaData for "org.apache.hadoop.hive.metastore.model.MStorageDescriptor" is specified with a foreign-key at class level yet no "table" is defined. All foreign-keys at this level must have a table that the FK goes to.
2024-10-16T20:46:13,668 WARN  org.apache.hadoop.hive.metastore.MetastoreDirectSqlUtils:79 - Failed to execute [select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ? ] with parameters [hive]
javax.jdo.JDODataStoreException: Error executing SQL query "select "FUNCS"."FUNC_ID" from "FUNCS" LEFT JOIN "DBS" ON "FUNCS"."DB_ID" = "DBS"."DB_ID" where "DBS"."CTLG_NAME" = ?".
	
I have fixed that as well.

adamdec0173 commented:
public IgnoreClosePrintStream(OutputStream out) {
        super(out, false, StandardCharsets.UTF_8);
    }

Finally, this is needed or NPE is being trown

massdosage commented on CHANGELOG.md:
```diff
@@ -3,14 +3,45 @@ All notable changes to this project will be documented in this file.
+- Updated `maven-source-plugin` from `3.2.0` to `3.3.1`.
+- Updated `maven-javadoc-plugin` from `3.2.0` to `3.10.1`.
+- Updated `maven-gpg-plugin` from `1.6` to `3.2.7`.
+- Removed com.google.common.base.Predicates in HiveRunnerExtension/StandaloneHiveRunner as it is no longer using in a new version of reflections library
```
```suggestion
- Removed `com.google.common.base.Predicates` in `HiveRunnerExtension`/`StandaloneHiveRunner` as it is no longer using in a new version of reflections library.
```

massdosage commented on CHANGELOG.md:
```diff
@@ -3,14 +3,45 @@ All notable changes to this project will be documented in this file.
+- Updated `maven-javadoc-plugin` from `3.2.0` to `3.10.1`.
+- Updated `maven-gpg-plugin` from `1.6` to `3.2.7`.
+- Removed com.google.common.base.Predicates in HiveRunnerExtension/StandaloneHiveRunner as it is no longer using in a new version of reflections library
+- Updated HiveConf property names in StandaloneHiveServerContext
```
```suggestion
- Updated `HiveConf` property names in `StandaloneHiveServerContext`.
```

massdosage commented on CHANGELOG.md:
```diff
@@ -3,14 +3,45 @@ All notable changes to this project will be documented in this file.
+- Updated `maven-gpg-plugin` from `1.6` to `3.2.7`.
+- Removed com.google.common.base.Predicates in HiveRunnerExtension/StandaloneHiveRunner as it is no longer using in a new version of reflections library
+- Updated HiveConf property names in StandaloneHiveServerContext
+- Set METASTORE_VALIDATE_CONSTRAINTS, METASTORE_VALIDATE_COLUMNS, METASTORE_VALIDATE_TABLES properties to false in StandaloneHiveServerContext
```
```suggestion
- Set `METASTORE_VALIDATE_CONSTRAINTS`, `METASTORE_VALIDATE_COLUMNS`, `METASTORE_VALIDATE_TABLES` properties to false in `StandaloneHiveServerContext`.
```

massdosage commented on CHANGELOG.md:
```diff
@@ -3,14 +3,45 @@ All notable changes to this project will be documented in this file.
+- Removed com.google.common.base.Predicates in HiveRunnerExtension/StandaloneHiveRunner as it is no longer using in a new version of reflections library
+- Updated HiveConf property names in StandaloneHiveServerContext
+- Set METASTORE_VALIDATE_CONSTRAINTS, METASTORE_VALIDATE_COLUMNS, METASTORE_VALIDATE_TABLES properties to false in StandaloneHiveServerContext
 - Added missing Hive & Datanucleus properties in StandaloneHiveServerContext so now the framework works on a new hive dependency versions
```
```suggestion
- Added missing Hive & Datanucleus properties in StandaloneHiveServerContext so now the framework works with new Hive dependency versions.
```

massdosage commented on src/test/java/com/klarna/hiverunner/examples/HelloHiveRunnerParamaterizedTest.java:
```diff
@@ -1,6 +1,6 @@
 /**
  * Copyright (C) 2013-2021 Klarna AB
- * Copyright (C) 2021-2024 The HiveRunner Contributors
+ * Copyright (C) 2021 The HiveRunner Contributors
```
Once this is merged I'll take a look at the copyright plugin and regenerate all the headers.

massdosage reviewed (changes requested):
This is looking good, just a few small comments from me and then we should be good to go to merge it.

massdosage commented on pom.xml:
```diff
@@ -33,155 +33,406 @@
+                    <groupId>commons-logging</groupId>
+                    <artifactId>commons-logging</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.apache.logging.log4j</groupId>
+                    <artifactId>log4j-slf4j-impl</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.apache.logging.log4j</groupId>
+                    <artifactId>log4j-slf4j</artifactId>
+                </exclusion>
```
Thoughts on replacing this with 
```
                <exclusion>
                    <groupId>org.apache.logging.log4j</groupId>
                    <artifactId>*</artifactId>
                </exclusion>
```
here and below?

massdosage commented on src/main/java/com/klarna/hiverunner/HiveServerContainer.java:
```diff
@@ -86,6 +86,7 @@ public void init(Map<String, String> testConfig, Map<String, String> hiveVars) {
         try {
             hiveServer2 = new HiveServer2();
             hiveServer2.init(hiveConf);
+            hiveServer2.start();
```
Just curious if this is the cause of those INFO logs that you mentioned over e-mail? I assume this call to `start` is required? Why wasn't it needed in the past?

adamdec0173 replied:
This exception is being thrown without it:

```
java.lang.IllegalStateException: Failed to create HiveServer :Cannot open sessions on an inactive HS2 instance, or the HS2 server leader is not ready; please use service discovery to connect the server leader again

	at com.klarna.hiverunner.HiveServerContainer.init(HiveServerContainer.java:106)
	at com.klarna.hiverunner.builder.HiveShellBase.start(HiveShellBase.java:160)
	at com.klarna.hiverunner.HiveRunnerCore.buildShell(HiveRunnerCore.java:87)
	at com.klarna.hiverunner.HiveRunnerCore.createHiveServerContainer(HiveRunnerCore.java:57)
	at com.klarna.hiverunner.HiveRunnerExtension.createHiveServerContainer(HiveRunnerExtension.java:107)
	at com.klarna.hiverunner.HiveRunnerExtension.postProcessTestInstance(HiveRunnerExtension.java:69)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: org.apache.hive.service.cli.HiveSQLException: Cannot open sessions on an inactive HS2 instance, or the HS2 server leader is not ready; please use service discovery to connect the server leader again
	at org.apache.hive.service.cli.session.SessionManager.createSession(SessionManager.java:436)
	at org.apache.hive.service.cli.session.SessionManager.openSession(SessionManager.java:425)
	at org.apache.hive.service.cli.CLIService.openSession(CLIService.java:211)
	at com.klarna.hiverunner.HiveServerContainer.init(HiveServerContainer.java:100)
	... 19 more
```

I have found this after a very long debugging of a new Hive code-base. I did not check the old one so can not compare.

adamdec0173 replied:
This exception is being thrown without it:

```
java.lang.IllegalStateException: Failed to create HiveServer :Cannot open sessions on an inactive HS2 instance, or the HS2 server leader is not ready; please use service discovery to connect the server leader again

	at com.klarna.hiverunner.HiveServerContainer.init(HiveServerContainer.java:106)
	at com.klarna.hiverunner.builder.HiveShellBase.start(HiveShellBase.java:160)
	at com.klarna.hiverunner.HiveRunnerCore.buildShell(HiveRunnerCore.java:87)
	at com.klarna.hiverunner.HiveRunnerCore.createHiveServerContainer(HiveRunnerCore.java:57)
	at com.klarna.hiverunner.HiveRunnerExtension.createHiveServerContainer(HiveRunnerExtension.java:107)
	at com.klarna.hiverunner.HiveRunnerExtension.postProcessTestInstance(HiveRunnerExtension.java:69)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at java.base/java.util.Optional.orElseGet(Optional.java:364)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
Caused by: org.apache.hive.service.cli.HiveSQLException: Cannot open sessions on an inactive HS2 instance, or the HS2 server leader is not ready; please use service discovery to connect the server leader again
	at org.apache.hive.service.cli.session.SessionManager.createSession(SessionManager.java:436)
	at org.apache.hive.service.cli.session.SessionManager.openSession(SessionManager.java:425)
	at org.apache.hive.service.cli.CLIService.openSession(CLIService.java:211)
	at com.klarna.hiverunner.HiveServerContainer.init(HiveServerContainer.java:100)
	... 19 more
```
I have found this after a very long debugging of a new Hive code-base. I did not check the old one so can not compare.

massdosage replied:
OK, so it sounds like we have to keep the above for now.

massdosage commented:
Thanks for the contribution @adamdec0173! I'll work on the license plugin in this branch and see what I can do about getting a release done, hopefully this week.

massdosage commented:
Fixes #180
